File: src/rest-secure.js
/*
 ----------------------------------------------------------------------------
 | REST-Secure: Secure REST Browser Interface                                |
 |                                                                           |
 | Copyright (c) 2022 M/Gateway Developments Ltd,                            |
 | Redhill, Surrey UK.                                                       |
 | All rights reserved.                                                      |
 |                                                                           |
 | http://www.mgateway.com                                                   |
 | Email: rtweed@mgateway.com                                                |
 |                                                                           |
 |                                                                           |
 | Licensed under the Apache License, Version 2.0 (the "License");           |
 | you may not use this file except in compliance with the License.          |
 | You may obtain a copy of the License at                                   |
 |                                                                           |
 |     http://www.apache.org/licenses/LICENSE-2.0                            |
 |                                                                           |
 | Unless required by applicable law or agreed to in writing, software       |
 | distributed under the License is distributed on an "AS IS" BASIS,         |
 | WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  |
 | See the License for the specific language governing permissions and       |
 |  limitations under the License.                                           |
 ----------------------------------------------------------------------------

26 August 2022 

 */

  // For source of these handler scripts, see the /idb_handlers folder

let handlerCode = new Map([
  ['instantiate', `
(function () {
  let initialised = false;
  let authorization = false;
  let timeout = 20000; // 20 seconds default fetch timeout

  function isEmpty(obj) {
    for (const key in obj) {
      return false;
    }
    return true;
  }

  function uuidv4(protocol) {
    if (protocol === 'https:') {
      return ([1e7]+-1e3+-4e3+-8e3+-1e11).replace(/[018]/g, c =>
        (c ^ crypto.getRandomValues(new Uint8Array(1))[0] & 15 >> c / 4).toString(16)
      );
    }
    else {
      return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function(c) {
        let r = Math.random() * 16 | 0, v = c == 'x' ? r : (r & 0x3 | 0x8);
        return v.toString(16);
      });
    }
  }

  function jwt_decode(jwt) {

    // Adapted from Auth0: https://github.com/auth0/jwt-decode

    let chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=";

    function InvalidCharacterError(message) {
      this.message = message;
    }

    InvalidCharacterError.prototype = new Error();
    InvalidCharacterError.prototype.name = "InvalidCharacterError";

    function b64DecodeUnicode(str) {
      return decodeURIComponent(
        atob(str).replace(/(.)/g, function(m, p) {
          var code = p.charCodeAt(0).toString(16).toUpperCase();
          if (code.length < 2) {
            code = "0" + code;
          }
          return "%" + code;
        })
      );
    }

    function base64_url_decode(str) {
      var output = str.replace(/-/g, "+").replace(/_/g, "/");
      switch (output.length % 4) {
        case 0:
          break;
        case 2:
          output += "==";
          break;
        case 3:
          output += "=";
          break;
        default:
          throw "Illegal base64url string!";
      }

      try {
        return b64DecodeUnicode(output);
      }
      catch (err) {
        return atob(output);
      }
    }

    function InvalidTokenError(message) {
      this.message = message;
    }

    InvalidTokenError.prototype = new Error();
    InvalidTokenError.prototype.name = "InvalidTokenError";

    function jwtDecode(token, options) {
      if (typeof token !== "string") {
        throw new InvalidTokenError("Invalid token specified");
      }
      options = options || {};
      var pos = options.header === true ? 0 : 1;
      try {
        return JSON.parse(base64_url_decode(token.split(".")[pos]));
      }
      catch (e) {
        throw new InvalidTokenError("Invalid token specified: " + e.message);
      }
    }

    return jwtDecode(jwt);
  }

  self.handler = async function(obj, finished) {

    // obj.endpoint
    // obj.protocol
    // obj.contentType
    // obj.useXAuth (true | false)
    // obj.responseHandlerUrl (optional)
    // obj.authTokenType (Bearer is default, or Token, empty string)
    // obj.refresh_token_uri
    // obj.timeout
    // obj.authHeader.request
    // obj.authHeader.response

    let worker = this;

    let loadRestResources = obj.load_rest_resources || false;
    let responseHandlerUrl = obj.responseHandlerUrl;
    let authTokenType = obj.authTokenType || 'Bearer ';
    if (authTokenType !== '' && authTokenType.slice(-1) !== ' ') authTokenType= authTokenType + ' ';
    let useJWT = obj.useJWT || false;
    let refresh_token_uri = obj.refresh_token_uri;
    let authHeader = obj.authHeader;
    if (!authHeader) authHeader = {};
    if (!authHeader.request) authHeader.request = 'Authorization';
    if (!authHeader.response) authHeader.response = 'authorization';

    if (obj.timeout) timeout = obj.timeout;
    let defaultTimeout = timeout;

    if (initialised) {
      return finished({
        error: 'QOper8 WebWorker has already been initiated'
      });
    }

    self.restSecure = {};

    self.restSecure.isInitialised = function() {
      return initialised;
    };

    let useXAuth = obj.useXAuth;

    if (typeof obj.endpoint === 'undefined') {
      return finished({
        error: 'No endpoint provided'
      });
    }

    let token = uuidv4();
    let endpoint = obj.endpoint;
    let contentType = obj.contentType || 'application/json';
    let jwt;

    let rest_resources = obj.rest_resources || {};
  
    self.restSecure.isValidToken = function(token_input) {
      return token_input === token;
    };

    self.restSecure.fetch = async function(opt) {

      const abortController = new AbortController();
      const abortTimer = setTimeout(function() {
        abortController.abort();
      }, timeout);

      // if the authorization token is a JWT, check to see if it's expired
      //  if it has, then request a new JWT using the refresh token (which is in the secure cookie)

      if (refresh_token_uri && jwt && jwt.exp < (Date.now()/1000)) {

        // fetch a new JWT using the refresh token

        let resource = endpoint + refresh_token_uri;
        let options = {
          method: 'GET',
          headers: {
            'Content-type': 'application/json'
          },
          signal: abortController.signal
        };

        try {
          let res = await fetch(resource, options);
          clearTimeout(abortTimer);
          let auth = res.headers.get(authHeader.response);
          if (auth && auth !== '') authorization = auth;

          // reset the fetch abort timer

          abortTimer = setTimeout(function() {
            abortController.abort()
          }, timeout);
        }
        catch(err) {
          // reduce the timeout for subsequent requests
          timeout = 1000;

          setTimeout(function() {
            // restore default timeout in 5 minutes time
            timeout = defaultTimeout;
          }, 300000);
        }
      }

      if (!opt.headers) opt.headers = {};
      let cType = opt.headers.contentType || contentType;

      if (cType === 'application/json' && typeof opt.body === 'object') {
        opt.body = JSON.stringify(opt.body);
      }

      let resource = endpoint + opt.uri;

      let restToken = uuidv4();
      let options = {
        method: opt.method,
        body: opt.body,
        headers: {
          'Content-type': cType,
          'X-Authorization': 'Bearer ' + restToken
        },
        signal: abortController.signal
      };

      if (opt.headers) {
        for (const name in opt.headers) {
          let namelc = name.toLowerCase();
          if (!['content-type', 'x-authorization', 'authorization'].includes(namelc)) {
            options.headers[name] = opt.headers[name];
          }
        }
      }

      if (authorization) {
        options.headers[authHeader.request] = authTokenType + authorization;
      }

      // **** Do the actual fetch now *****

      let response;
      try {
        response = await fetch(resource, options);
        clearTimeout(abortTimer);
      }
      catch(err) {
        timeout = 1000;

        setTimeout(function() {
          // restore default timeout in 5 minutes time
          timeout = defaultTimeout;
        }, 300000);

        return {
          body: {
            error: 'Request timed out'
          }
        };
      }

      let headers = {};
      for (const [key, value] of response.headers.entries()) {
        headers[key] = value;
      }
      if (useXAuth) {
        let returnedToken = response.headers.get('x-request-id');
        if (restToken !== returnedToken) {
          return {error: 'Unable to match x-authorization token'};
        }
      }
      if (cType === 'application/json') {
        response = await response.json();
      }

      if (opt.responseHandlerUrl) {
        importScripts(opt.responseHandlerUrl);
        if (self.responseHandler) {
          let resObj = self.responseHandler(response, headers);
          response = resObj.response;
          headers = resObj.headers;
        }
      }

      if (headers[authHeader.response]) {
        authorization = headers[authHeader.response];
        delete headers[authHeader.response];

        // is this a JWT?

        try {
          jwt = jwt_decode(authorization);
          if (opt.jwt_claims) {
            opt.jwt_claims.forEach(function(claim) {
              response[claim] = jwt[claim];
            });
          }
        }
        catch(err) {
        }
      }

      return {
        body: response,
        headers: headers
      };
    };

    let already_authenticated = false;
    let claims;
    if (refresh_token_uri) {

      // when starting up for the first time, see if there's a valid
      // refresh token in the browser's secure cookie.  If so, fetch and
      // save a fresh JWT or idToken


      let res = await self.restSecure.fetch({
        uri: refresh_token_uri
      });

      claims = res.body && res.body.claims;

      if (authorization) already_authenticated = true;
    }

    initialised = true;
    finished({
      already_authenticated: already_authenticated,
      claims: claims,
      qoper8: {
        token: token
      }
    });
  };
})();

  `],

  ['secure_fetch', `
  self.handler = async function(obj, finished) {
    let worker = this;

    if (!self.restSecure || !self.restSecure.isInitialised()) {
      return finished({
        error: 'REST-Secure QOper8 WebWorker has not been initialised'
      });
    }

    if (!self.restSecure.isValidToken(obj.qoper8.token)) {
      return finished({
        error: 'Invalid token received from main process'
      });
    }

    let options = obj.options;

    let res = await self.restSecure.fetch({
      uri: obj.uri,
      method: options.method,
      body: options.body,
      headers: options.headers,
      responseHandlerUrl: options.responseHandlerUrl,
      jwt_claims: options.jwt_claims
    });

    if (res.headers) delete res.headers['x-request-id'];
    finished(res);
  };
  `]

]);

// *****  REST-Secure ********

let REST_Secure = class {
  constructor(options) {
    let QOper8 = options.QOper8;
    let qOptions = options.qOptions || {};

    this.name = 'REST-Secure';
    this.build = '0.1';
    this.buildDate = '26 August 2022';

    let qoper8 = new QOper8({
      poolSize: 1,
      logging: qOptions.logging,
      workerInactivityCheckInterval: qOptions.workerInactivityCheckInterval || 60,
      workerInactivityLimit: qOptions.workerInactivityLimit || 60
    });

    for (const [key, value] of handlerCode) {
      qoper8.handlersByMessageType.set(key, qoper8.createUrl(value));
    }

    let rs = this;

    this.start = async function() {

      let msg = {
        type: 'instantiate',
        contentType: options.contentType,
        protocol: window.location.protocol,
        endpoint: options.endpoint,
        authTokenType: options.authTokenType,
        refresh_token_uri: options.refresh_token_uri,
        timeout: options.timeout,
        authHeader: options.authHeader
      };

      let res = await qoper8.send(msg);

      if (res.error) {
        return res;
      }

      let already_authenticated = res.already_authenticated;
      let claims = res.claims;

      async function secure_fetch(uri, options) {

        let msg = {
          type: 'secure_fetch',
          uri: uri,
          options: options,
          qoper8: {
            token: res.qoper8.token
          }
        }
        return new Promise((resolve) => {
          qoper8.message(msg, function(responseObj) {
            resolve(responseObj);
          });
        });
      }

      rs.fetch = async function(uri, options) {
        options = options || {};
        if (options.responseHandlerCode) {
          options.responseHandlerUrl = qoper8.createUrl(options.responseHandlerCode);
        }
        let resp = await secure_fetch(uri, options);
        delete resp.qoper8;
        return resp;
      };

      return {
        already_authenticated: already_authenticated,
        claims: claims
      };
    }
  }
};

export {REST_Secure};

File: mgweb/RestSecure.m
RestSecure ; mgweb-server Back-end for interoperability with Rest-Secure-based front end
 ;
 ;----------------------------------------------------------------------------
 ;| RestSecure: mgweb-server Interface for Rest-Secure applications          |
 ;|                                                                          |
 ;| Copyright (c) 2022 M/Gateway Developments Ltd,                           |
 ;| Redhill, Surrey UK.                                                      |
 ;| All rights reserved.                                                     |
 ;|                                                                          |
 ;| http://www.mgateway.com                                                  |
 ;| Email: rtweed@mgateway.com                                               |
 ;|                                                                          |
 ;|                                                                          |
 ;| Licensed under the Apache License, Version 2.0 (the "License");          |
 ;| you may not use this file except in compliance with the License.         |
 ;| You may obtain a copy of the License at                                  |
 ;|                                                                          |
 ;|     http://www.apache.org/licenses/LICENSE-2.0                           |
 ;|                                                                          |
 ;| Unless required by applicable law or agreed to in writing, software      |
 ;| distributed under the License is distributed on an "AS IS" BASIS,        |
 ;| WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. |
 ;| See the License for the specific language governing permissions and      |
 ;|  limitations under the License.                                          |
 ;----------------------------------------------------------------------------
 ;
 ; 17 August 2022
 ;
 QUIT
 ;
deleteExpiredIdTokens
 ;
 n expires,idToken,now,uuid
 ;
 ; Expired Id and Refresh Tokens
 ;
 l +^RestSecure
 s now=$$convertDateToSeconds($h)
 s uuid=""
 f  s uuid=$o(^RestSecure("refreshToken",uuid)) q:uuid=""  d
 . s expires=$g(^RestSecure("refreshToken",uuid,"expires"))
 . i expires<now d
 . . s idToken=$g(^RestSecure("refreshToken",uuid,"idToken"))
 . . i idToken'="" k ^RestSecure("idToken",idToken)
 . . k ^RestSecure("refreshToken",uuid)
 ;
 ; remove stranded idTokens
 ;
 s uuid=""
 f  s uuid=$o(^RestSecure("idToken",uuid)) q:uuid=""  d
 . s refreshToken=$g(^RestSecure("idToken",uuid,"refreshToken"))
 . i refreshToken'="",'$d(^RestSecure("refreshToken",refreshToken)) k ^RestSecure("idToken",uuid)
 l -^RestSecure
 QUIT
 ;
returnAuthToken(req,headers) ;
 ;
 n token
 ;
 s token=$g(req("headers","x_authorization"))
 i token="" QUIT
 s token=$p(token,"Bearer ",2)
 i token="" QUIT
 s headers("X-Request-ID")=token
 QUIT
 ;
getAuthToken(req) ;
 ;
 n token
 ;
 s token=$g(req("headers","x_authorization"))
 i token="" QUIT ""
 s token=$p(token,"Bearer ",2)
 QUIT token
 ;
getAuthorization(req) ;
 ;
 n token
 ;
 s token=$g(req("headers","authorization"))
 i token="" QUIT ""
 s token=$p(token,"Bearer ",2)
 QUIT token
 ;
deleteExpiredJWTTokens
 ;
 n expires,now,refreshToken,uuid
 ;
 l +^RestSecure
 s now=$$convertDateToSeconds($h)
 s uuid=""
 f  s uuid=$o(^RestSecure("jwt",uuid)) q:uuid=""  d
 . s refreshToken=$g(^RestSecure("jwt",uuid,"refreshToken"))
 . s expires=$g(^RestSecure("jwt",uuid,"expires"))
 . i expires<now k ^RestSecure("jwt",uuid)
 s refreshToken=""
 f  s refreshToken=$o(^RestSecure("refreshToken",refreshToken)) q:refreshToken=""  d
 . s expires=$g(^RestSecure("refreshToken",refreshToken,"expires"))
 . i expires<now k ^RestSecure("refreshToken",refreshToken)
 l -^RestSecure
 QUIT
 ;
getRefreshToken(req,data) ;
 ;
 k data
 n id,token
 ;
 d deleteExpiredJWTTokens
 s token=$g(req("headers","cookie"))
 i token="" QUIT 0
 s token=$p(token,"RefreshToken=",2)
 i token="" QUIT 0
 s id=$g(^RestSecure("refreshToken",token,"id"))
 i id="" QUIT 0
 s data("refreshToken")=token
 s data("id")=id
 QUIT 1
 ;
validJWT(req,errors,claims) ;
 ;
 n key,jwt,rsjwt,uuid
 ;
 k claims
 d deleteExpiredJWTTokens
 ;
 ; was a jwt returned as the authorization header?
 ;
 s jwt=$$getAuthorization(.req)
 i jwt="" d  QUIT 0
 . s errors("error")="Missing Authorization Header"
 ;
 ; does the jwt have a uuid to link it to back-end copy?
 ;
 s uuid=$$getClaim^%zmgwebJWT(jwt,"uuid")
 i uuid="" d  QUIT 0
 . s errors("error")="Invalid JWT 1"
 ;
 ; is there a back-end copy jwt with that UUID?
 ;
 s rsjwt=$g(^RestSecure("jwt",uuid,"jwt"))
 i rsjwt="" d  QUIT 0
 . s errors("error")="Invalid or expired JWT"
 ;
 ; do the JWTs match?
 ;
 i jwt'=rsjwt d  QUIT 0
 . s errors("error")="Invalid JWT 3"
 ;
 ; retrieve the key
 ;
 s key=$g(^RestSecure("jwt",uuid,"key"))
 i key="" d  QUIT 0
 . s errors("error")="Invalid or Expired JWT 2"
 ;
 ; authenticate the incoming JWT using the key
 ;
 i '$$authenticateJWT^%zmgwebJWT(jwt,key,.reason) d  QUIT 0
 . s errors("error")=reason
 . s ^rob("update","reason")=reason
 ;
 ; JWT is OK!
 ;
 m claims=^RestSecure("jwt",uuid,"claims")
 s claims("id")=^RestSecure("jwt",uuid,"id")
 s claims("uuid")=uuid
 ;
 QUIT 1
 ;
createJWT(id,claims,options,headers) ;
 ;
 n claim,field,jwt,key,now,refreshToken,sameSite,secure,uuid
 ;
 i '$g(options("jwtExpiryTime")) s options("jwtExpiryTime")=300
 i '$g(options("refreshTokenExpiryTime")) s options("refreshTokenExpiryTime")=86400
 s secure="; Secure"
 i $g(options("secureCookie"))=0 secure=""
 s sameSite="; SameSite=Strict"
 i $g(options("sameSiteCookie"))=0 sameSite=""
 d deleteExpiredJWTTokens
 s key=$$createJWTUid^%zmgwebJWT()
 s uuid=$$createJWTUid^%zmgwebJWT()
 s refreshToken=$$createJWTUid^%zmgwebJWT()
 s now=$$convertDateToSeconds($h)
 ;
 l +^RestSecure
 s ^RestSecure("refreshToken",refreshToken,"expires")=now+options("refreshTokenExpiryTime")
 s ^RestSecure("refreshToken",refreshToken,"id")=$g(id)
 s ^RestSecure("refreshToken",refreshToken,"jwtExpiryTime")=options("jwtExpiryTime")
 s ^RestSecure("jwt",uuid,"key")=key
 s ^RestSecure("jwt",uuid,"id")=id
 s ^RestSecure("jwt",uuid,"created")=now
 s ^RestSecure("jwt",uuid,"expires")=now+options("jwtExpiryTime")
 s ^RestSecure("jwt",uuid,"refreshToken")=refreshToken
 s claim=""
 f  s claim=$o(claims(claim)) q:claim=""  d
 . n claimValue
 . s claimValue=$g(claims(claim))
 . s ^RestSecure("jwt",uuid,"claims",claim)=claimValue
 s claims("uuid")=uuid
 s claims("iss")="RestSecure"
 s jwt=$$createJWT^%zmgwebJWT(.claims,options("jwtExpiryTime"),key)
 s ^RestSecure("jwt",uuid,"jwt")=jwt
 l -^RestSecure
 ;
 s headers("Authorization")=jwt
 s headers("Set-Cookie")="RefreshToken="_refreshToken_"; Path=/; HttpOnly; Max-Age="_options("refreshTokenExpiryTime")_secure_sameSite_";"
 QUIT
 ;
refreshJWT(id,refreshToken,inClaims,headers) ;
 ;
 n claim,claims,expiryTime,field,jwt,key,now,uuid
 ;
 m claims=inClaims
 d deleteExpiredJWTTokens
 s key=$$createJWTUid^%zmgwebJWT()
 s uuid=$$createJWTUid^%zmgwebJWT()
 s now=$$convertDateToSeconds($h)
 ;
 s expiryTime=$g(^RestSecure("refreshToken",refreshToken,"jwtExpiryTime"))
 i expiryTime="" s expiryTime=300
 ;
 l +^RestSecure
 s ^RestSecure("jwt",uuid,"key")=key
 s ^RestSecure("jwt",uuid,"id")=id
 s ^RestSecure("jwt",uuid,"created")=now
 s ^RestSecure("jwt",uuid,"expires")=now+expiryTime
 s ^RestSecure("jwt",uuid,"refreshToken")=$g(refreshToken)
 s claim=""
 f  s claim=$o(claims(claim)) q:claim=""  d
 . n claimValue
 . s claimValue=$g(claims(claim))
 . s ^RestSecure("jwt",uuid,"claims",claim)=claimValue
 s claims("uuid")=uuid
 s claims("iss")="RestSecure"
 s jwt=$$createJWT^%zmgwebJWT(.claims,expiryTime,key)
 s ^RestSecure("jwt",uuid,"jwt")=jwt
 l -^RestSecure
 ;
 s headers("Authorization")=jwt
 QUIT
 ;
refreshIdToken(id,refreshToken,headers) ;
 ;
 n uuid
 ;
 d deleteExpiredIdTokens
 s uuid=$$createJWTUid^%zmgwebJWT()
 ;
 l +^RestSecure
 s ^RestSecure("idToken",uuid,"refreshToken")=refreshToken
 s ^RestSecure("refreshToken",refreshToken,"idToken")=uuid
 l -^RestSecure
 ;
 s headers("Authorization")=uuid
 QUIT
 ;
updateJWT(req,res,headers,errors,expiryTime) ;
 ;
 n claims,jwt,key,now,uuid
 ;
 s jwt=$$getAuthorization(.req)
 i jwt="" d  QUIT
 . s errors("error")="Missing Authorization Header"
 s ^rob("update","jwt")=jwt
 s uuid=$$getClaim^%zmgwebJWT(jwt,"uuid")
 i uuid="" d  QUIT 0
 . s errors("error")="Invalid JWT 1"
 s ^rob("update","uuid")=uuid
 i '$g(expiryTime) s expiryTime=86400
 d deleteExpiredJWTTokens
 i '$d(^RestSecure("jwt",uuid)) d  QUIT 0
 . s errors("error")="Invalid or Expired JWT 1"
 i jwt'=$g(^RestSecure("jwt",uuid,"jwt")) d  QUIT 0
 . s errors("error")="Invalid JWT 2"
 s key=$g(^RestSecure("jwt",uuid,"key"))
 i key="" d  QUIT 0
 . s errors("error")="Invalid or Expired JWT 2"
 s ^rob("update","key")=key
 i '$$authenticateJWT^%zmgwebJWT(jwt,key,.reason) d  QUIT 0
 . s errors("error")=reason
 . s ^rob("update","reason")=reason
 ;
 s now=$$convertDateToSeconds($h)
 i now'=$g(^RestSecure("jwt",uuid,"now")) d
 . l +^RestSecure
 . s ^RestSecure("jwt",uuid,"expires")=now+expiryTime
 . m claims=req("jwt_claims")
 . s claims("uuid")=uuid
 . s claims("iss")="RestSecure"
 . s jwt=$$createJWT^%zmgwebJWT(.claims,expiryTime,key)
 . s ^RestSecure("jwt",uuid,"jwt")=jwt
 . l -^RestSecure
 s headers("Authorization")=jwt
 d returnAuthToken^RestSecure(.req,.headers)
 QUIT 1
 ;
createIdToken(id,options,headers) ;
 ;
 n now,refreshToken,sameSite,secure,uuid
 ;
 i '$g(options("refreshTokenExpiryTime")) s options("refreshTokenExpiryTime")=86400
 s secure="; Secure"
 i $g(options("secureCookie"))=0 secure=""
 s sameSite="; SameSite=Strict"
 i $g(options("sameSiteCookie"))=0 sameSite=""
 d deleteExpiredIdTokens
 s uuid=$$createJWTUid^%zmgwebJWT()
 s now=$$convertDateToSeconds($h)
 s refreshToken=$$createJWTUid^%zmgwebJWT()
 ;
 l +^RestSecure
 s ^RestSecure("refreshToken",refreshToken,"expires")=now+options("refreshTokenExpiryTime")
 s ^RestSecure("refreshToken",refreshToken,"expiryTime")=options("refreshTokenExpiryTime")
 s ^RestSecure("refreshToken",refreshToken,"id")=$g(id)
 s ^RestSecure("refreshToken",refreshToken,"idToken")=uuid
 ;
 s ^RestSecure("idToken",uuid,"refreshToken")=refreshToken
 l -^RestSecure
 ;
 s headers("Authorization")=uuid
 s headers("Set-Cookie")="RefreshToken="_refreshToken_"; Path=/; HttpOnly; Max-Age="_options("refreshTokenExpiryTime")_secure_sameSite_";"
 QUIT
 ;
validIdToken(req,errors,data) ;
 ;
 n error,idToken,key,refreshToken
 ;
 k data
 d deleteExpiredIdTokens
 ;
 ; was an idToken returned as the authorization header?
 ;
 s error=0
 s idToken=$$getAuthorization(.req)
 i idToken="" d  i error QUIT 0
 . ; 
 . ; See if the RefreshToken cookie is valid - if so, allow this through
 . ; by creating a new temporary idToken
 . ;
 . n crt,newUuid
 . s cookie=$g(req("headers","cookie"))
 . i cookie="" d  QUIT 
 . . s errors("error")="No cookie"
 . . s error=1
 . ;
 . s crt=$p(cookie,"RefreshToken=",2)
 . i crt="" d  QUIT
 . . s errors("error")="Invalid cookie"
 . . s error=1
 . i '$d(^RestSecure("refreshToken",crt) d  QUIT
 . . s errors("error")="Unrecognised cookie"
 . . s error=1
 . s newUuid=$$createJWTUid^%zmgwebJWT()
 . l +^RestSecure
 . s ^RestSecure("idToken",newUuid,"refreshToken")=crt
 . l -^RestSecure
 . idToken=newUuid
 ;
 ; is there a back-end copy idToken with the same UUID?
 ;
 i '$d(^RestSecure("idToken",idToken)) d  QUIT 0
 . s errors("error")="Invalid or expired authorization token"
 ;
 ; idToken is OK!
 ;
 s refreshToken=$g(^RestSecure("idToken",idToken,"refreshToken"))
 i refreshToken="" d  QUIT 0
 . s errors("error")="Invalid or expired authorization token"
 ;
 s data("id")=$g(^RestSecure("refreshToken",refreshToken,"id"))
 s data("idToken")=idToken
 s data("refreshToken")=refreshToken
 ;
 QUIT 1
 ;
updateIdToken(idToken,refreshToken,headers,errors)
 ;
 n expiryTime,newUuid
 ;
 i idToken="" d  QUIT
 . s errors("error")="Missing idToken"
 i refreshToken="" d  QUIT 0
 . s errors("error")="Missing Refresh Token"
 s newUuid=$$createJWTUid^%zmgwebJWT()
 ;
 l +^RestSecure
 k ^RestSecure("idToken",idToken);
 s ^RestSecure("idToken",newUuid,"refreshToken")=refreshToken
 s ^RestSecure("refreshToken",refreshToken,"idToken")=newUuid
 l -^RestSecure
 ;
 s headers("Authorization")=newUuid
 QUIT
 ;
convertDateToSeconds(hdate)
 Q (hdate*86400)+$p(hdate,",",2)
 ;